sql/sql_insert.cc

```diff
@@ -1645,6 +1645,12 @@ int write_record(THD *thd, TABLE *table,COPY_INFO *info)
 	error= HA_ERR_FOUND_DUPP_KEY;         /* Database can't find key */
 	goto err;
       }
+	  if (table->key_info[key_nr].algorithm == HA_KEY_ALG_HASH &&
+       strcmp(table->file->table_type(), "InnoDB") == 0)
```

Comment:
Here is what you should do instead using db_type.

sql/handler.h

add 

/* 
  Storage engine supports unique indexes to blob columns.
*/
#define HA_CAN_UNIQUE_BLOB_INDEXES          (1LL << 47)

In storage/xtradb/handler/ha_innodb.cc constructor ha_innobase::ha_innobase(

add HA_CAN_UNIQUE_BLOB_INDEXES,

Find similar place from MyISAM. Then here you should do table->file->ha_table_flags() & HA_CAN_UNIQUE_BLOB_INDEXES.